source/extensionPoints.py

```diff
@@ -0,0 +1,203 @@
+	and you call:
+		callWithSupportedKwargs(myFunc, a=1, b=2, c=3)
+	Instead of raising a TypeError, myFunc will simply be called like this:
+		myFunc(a=1, b=2)
+	"""
+	spec = inspect.getargspec(func)
+	if spec.keywords:
+		# func has a catch-all for kwargs (**kwargs).
+		return func(*args, **kwargs)
+	# spec.defaults lists all arguments with defaults (keyword arguments).
+	numKwargs = len(spec.defaults) if spec.defaults else 0
+	# spec.args lists all arguments, first positional, then keyword.
+	firstKwarg = len(spec.args) - numKwargs
+	supportedKwargs = set(spec.args[firstKwarg:])
+	for kwarg in kwargs.keys():
+		if kwarg not in supportedKwargs:
+			del kwargs[kwarg]
+	return func(*args, **kwargs)
+
+class Action(HandlerRegistrar):
+	"""Allows interested parties to register to be notified when some action occurs.
+	For example, this might be used to notify that the configuration profile has been switched.
+
+	First, an Action is created:
+
+	>>> somethingHappened = extensionPoints.Action()
+
+	Interested parties then register to be notified about this action:
+
+	>>> def onSomethingHappened(someArg=None):
+		... 	print(someArg)
+	...
+	>>> somethingHappened.register(onSomethingHappened)
+
+	When the action is performed, register handlers are notified:
+
+	>>> somethingHappened.notify(someArg=42)
+	"""
+
+	def notify(self, **kwargs):
+		"""Notify all registered handlers that the action has occurred.
+		@param kwargs: Arguments to pass to the handlers.
+		"""
+		for handler in self.handlers:
+			try:
+				callWithSupportedKwargs(handler, **kwargs)
+			except:
+				log.exception("Error running handler %r for %r" % (handler, self))
+
+class Filter(HandlerRegistrar):
+	"""Allows interested parties to register to modify a specific kind of data.
+	For example, this might be used to allow modification of spoken messages before they are passed to the synthesizer.
+
+	First, a Filter is created:
+
+	>>> messageFilter = extensionPoints.Filter()
+
+	Interested parties then register to filter the data:
+
+	>>> def filterMessage(message, someArg=None):
+	... 	return message + " which has been filtered."
+	...
+	>>> messageFilter.register(filterMessage)
+
+	When filtering is desired, all registered handlers are called to filter the data:
+
+	>>> messageFilter.apply("This is a message", someArg=42)
+	'This is a message which has been filtered'
+	"""
+
+	def apply(self, value, **kwargs):
+		"""Pass a value to be filtered through all registered handlers.
+		The value is passed to the first handler
+		and the return value from that handler is passed to the next handler.
+		This process continues for all handlers until the final handler.
+		The return value from the final handler is returned to the caller.
+		@param value: The value to be filtered.
+		@param kwargs: Arguments to pass to the handlers.
+		@return: The filtered value.
+		"""
+		for handler in self.handlers:
+			try:
+				value = callWithSupportedKwargs(handler, value, **kwargs)
+			except:

```

Comment:
Filters shouldn't explicitly depend on each other. That just isn't possible in an ecosystem where you can't know what filters will be installed. Of course an earlier filter affects a later filter by changing the value, but the later filter should not depend on specific behaviour provided by a particular earlier filter. For all you know, the earlier filter didn't get registered or a filter got registered in between them. If you want that kind of explicit dependency, the correct solution is to use two filters; e.g. a pre-process filter and a post-process filter.